lxd/storage/backend_lxd.go

```diff
@@ -320,8 +326,146 @@ func (b *lxdBackend) CreateInstanceFromBackup(inst Instance, sourcePath string,
+	}
+
+	volType, err := InstanceTypeToVolumeType(inst.Type())
+	if err != nil {
+		return err
+	}
+
+	volDBType, err := VolumeTypeToDBType(volType)
+	if err != nil {
+		return err
+	}
+
+	contentType := drivers.ContentTypeFS
+	if inst.Type() == instancetype.VM {
+		contentType = drivers.ContentTypeBlock
+	}
+
+	// Get the root disk device config.
+	_, rootDiskConf, err := shared.GetRootDiskDevice(inst.ExpandedDevices().CloneNative())
+	if err != nil {
+		return err
+	}
+
+	// Initialise a new volume containing the root disk config supplied in the new instance.
+	vol := b.newVolume(volType, contentType, project.Prefix(inst.Project(), inst.Name()), rootDiskConf)
+
+	// We don't need to use the source instance's root disk config, so set to nil.
+	srcVol := b.newVolume(volType, contentType, project.Prefix(src.Project(), src.Name()), nil)
+
+	revert := true
+	defer func() {
+		if !revert {
+			return
+		}
+		b.DeleteInstance(inst, op)
+	}()
+
+	srcPool, err := GetPoolByInstance(b.state, src)
+	if err != nil {
+		return err
+	}
+
+	if b.Name() == srcPool.Name() {
+		logger.Debug("CreateInstanceFromCopy same-pool mode detected")
+		err = b.driver.CreateVolumeFromCopy(vol, srcVol, snapshots, op)
+		if err != nil {
+			return err
+		}
+	} else {
+		// We are copying volumes between storage pools so use migration system as it will
+		// be able to negotiate a common transfer method between pool types.
+		logger.Debug("CreateInstanceFromCopy cross-pool mode detected")
+
+		// If we are copying snapshots, retrieve a list of snapshots from source volume.
+		snapshotNames := []string{}
+		if snapshots {
+			snapshots, err := VolumeSnapshotsGet(b.state, srcPool.Name(), src.Name(), volDBType)
+			if err != nil {
+				return err
+			}
+
+			for _, snapshot := range snapshots {
+				_, snapShotName, _ := shared.ContainerGetParentAndSnapshotName(snapshot.Name)
+				snapshotNames = append(snapshotNames, snapShotName)
+			}
+		}
+
+		// Use in-memory pipe pair to simulate a connection between the sender and receiver.
+		aEnd, bEnd := memorypipe.NewPipePair()
+
+		// Negotiate the migration type to use.
+		offeredTypes := srcPool.MigrationTypes(contentType)
+		offerHeader := migration.TypesToHeader(offeredTypes...)
+		migrationType, err := migration.MatchTypes(offerHeader, migration.MigrationFSType_RSYNC, b.MigrationTypes(contentType))
+		if err != nil {
+			return fmt.Errorf("Failed to negotiate copy migration type: %v", err)
+		}
+
+		// Run sender and receiver in separate go routines to prevent deadlocks.
+		aEndErrCh := make(chan error, 1)
+		bEndErrCh := make(chan error, 1)
+		go func() {
+			err := srcPool.MigrateInstance(src, aEnd, migration.VolumeSourceArgs{
+				Name:          src.Name(),
+				Snapshots:     snapshotNames,
+				MigrationType: migrationType,
+				TrackProgress: true, // Do use a progress tracker on sender.
+			}, op)
+
+			aEndErrCh <- err
+		}()
+
+		go func() {
+			err := b.CreateInstanceFromMigration(inst, bEnd, migration.VolumeTargetArgs{
```

Comment:
@stgraber something I wanted to draw attention to here, the `CreateInstanceFromMigration` uses the supplied `insts`'s root device config when passing to the driver's `CreateVolumeFromMigration` function. This means that it uses the same root device config when re-creating all of the snapshots too. Is this acceptable/correct or should we be using the snapshotted root device config at the time the snapshot was taken?

Reply:
I think that's correct as snapshots are immutable yet the container and its snapshots can be moved between pools, so using the data from the snapshot may be incorrect.